### PR TITLE
Datasets

### LINKED ISSUE #5: Add Meta-Dataset and VTAB
## Is your feature request related to a problem? Please describe.

The current set of included datasets are too constrained, limited in diversity, or only weakly related to representation quality. [Meta-Dataset (MD)][md] and the [Visual Task Adaptation Benchmark][vtab] define good representations as those that adapt to diverse, unseen tasks with few examples. These datasets really push all current datasets provided by this library into the toy domain.

Implementing this feature will allow for rapidly testing and comparing models and confidently say which will perform better in real world situations.
 
## Describe the solution you'd like

The [google-research/meta-dataset repostiroy][md-gh] is [Apache License Version 2.0][md-gh-license] which is compatible with our [MIT License][tmetal-license] so the quickest path to adding this feature will likely be to integrate the code in a way that is compatible with this libraries abstractions.

The biggest problem will be that the code (being from google) uses TensorFlow and is concerned with Python 2 compatibility which we do not care about supporting (only Python 3.6+). From [their installation instructions][md-gh-install]:

> Meta-Dataset is generally compatible with Python 2 and Python 3, but some parts of the code may require Python 3. The code works with TensorFlow 2, although it makes extensive use of tf.compat.v1 internally.

It looks like the code makes use of [TFRecords](https://www.tensorflow.org/api_docs/python/tf/data/TFRecordDataset) which will need to be removed.

## Describe alternatives you've considered

If migrating that code is not worth the effort the datasets required to create [Meta-Dataset][md] can be used from existing PyTorch implementations already in this dataset, [TorchVision Datasets](https://pytorch.org/vision/0.8/datasets.html), or other PyTorch libraries. The dataloader and/or dataset combining logic will likely still be of interest at least as a reference implementation.

I have also considered developing a novel dataset, but the time to add this dataset will add significant value in a much shorter timeframe so that should be revisited later.

## Behavior Changes

This feature is purely additive to the library and should not change behavior of any existing code or make any breaking API changes.

## Additional context

- [Meta-Dataset (MD) Paper][md]
- [Visual Task Adaptation Benchmark (VTAB) Paper][vtab]
- [MD and VTAB GitHub Repository reference implementation][md-gh]

[md]: https://arxiv.org/abs/1903.03096
[vtab]: https://arxiv.org/abs/1910.04867
[md-gh]: https://github.com/google-research/meta-dataset
[md-gh-license]: https://github.com/google-research/meta-dataset/blob/801bbe633540d4b9442252415f9334929c1608c0/LICENSE
[md-gh-install]: https://github.com/google-research/meta-dataset#installation
[tmetal-license]: https://github.com/sevro/torchmetal/blob/7b31204d1d3283a0966ec7b70698ea8a9c6b0341/LICENSE

### LINKED ISSUE #16: Review metadatset reader
Currently the implementation of metadataset in [Iron-beagle/torchmetal](https://github.com/iron-beagle/torchmetal/) on branch `datasets` does not regard the code of the [reader](https://github.com/google-research/meta-dataset/blob/main/meta_dataset/data/reader.py) in metadataset. This reader appears to have functionality that will be needed. Specifically in the `episode_representation_generator`. Both the `run_counter` and the logic regarding the `remaining` variable seem to indicate behavior that needs to be mimicked outside the already implemented sampler

### LINKED ISSUE #17: Padding Speed / Pad Size
Currently the Padding for batch size is done using a torch collate function in [ironbeagle-torchmetal](https://github.com/iron-beagle/torchmetal/blob/datasets/torchmetal/utils/data/dataloader.py) `BatchMetaCollatePad`. Its `call` hard-codes the largest batch size, `500` currently. Also, the only tested image shape has been concrete through transforms in the [helpers](https://github.com/iron-beagle/torchmetal/blob/datasets/torchmetal/datasets/helpers.py).

The desired behavior is to Pad according to the largest batch and largest image efficiently. Finding this across all tasks in an episode needs to be implemented and used in the `BatchMetaCollatePad`

### PR DESCRIPTION
**Pull request by iron-beagle**
## Changes

This pull will add the functionality outlined in [this feature request](https://github.com/emerald-ai/torchmetal/issues/5#issue-936019651)

## Checklist

- [ ] Lint and unit tests work locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

I have decided to replace the TensorFlow records with h5 files. I have changed how they are referenced slightly so that there is only one h5 file that needs to be opened for a dataset instead of an h5 file for each class. Other than this the dataset conversion works pretty much the same. My plan moving forward is to implement the Meta-Dataset sampler code that creates the [episodes](https://github.com/iron-beagle/meta-dataset/blob/7bbe8995e464eba8cfb5826bc9d6467ec26d8432/meta_dataset/data/sampling.py#L206) into what torchmetal calls a [splitter](https://github.com/iron-beagle/torchmetal/blob/7b31204d1d3283a0966ec7b70698ea8a9c6b0341/torchmeta/transforms/splitters.py).

### CODE REVIEW

**Comment by iron-beagle** [comment]
# Issues
commit: fb4b92e
## Transforms
Target Transforms that require num_classes can not be used with metadata's sampler as num_classes is variable and the same transform object is shared across all datasets being sampled from. Furthermore the splitter transform, which decides image indices for a class which is split into support and query, is also shared. Although I have not verified if it will, it is likely to cause concurrency issues if multiple workers sample tasks together. I suggest simply modifying transforms to be more functional to solve this issue

NOTE: Some of the utils functions in the transforms folder exist in the original dataset objects code.

## Sampler
The config for metadata's sampler is hardcoded right now. A way to pass in the settings needs to be implemented. Also the RNG is global hardcoded in the sampler.py and should be seeded. The metadata split object(enum) for the sampler should be linked somehow to torchmeta's meta_split in the datasets, right now it is also hardcoded in the sampler. 

## Prospected
There are errors on target shape using example code below. This could be resulting from the target transform issue above or something I missed in the dataloader/collate_function code. Most likely the lack of target transform as I think that is the only thing missing from the original example provided. Also the input shape is wrong for batch size.


`

    from torchvision import transforms
    from torchmetal.datasets import Omniglot
    from torchmetal.datasets.helpers import omniglot
    from torchmetal.transforms import ClassSplitter
    from torchmetal.utils.data import BatchMetaDataLoader, MetaDataSampler

    transform_split = ClassSplitter(variable_class_split=True)
    transform_tens = transforms.Compose([transforms.ToTensor()])

    dataset = Omniglot("data", num_classes_per_task=5, use_vinyals_split=False,
                    meta_dataset=True, transform=transform_tens,
                    dataset_transform=transform_split, download=True)

    sampler = MetaDataSampler(dataset)


    dataloader = BatchMetaDataLoader(dataset, shuffle=False, batch_size=16, num_workers=4,
                                    sampler=sampler)

    for batch in dataloader:
        train_inputs, train_targets = batch["train"]
        print('Train inputs shape: {0}'.format(train_inputs.shape))
        print('Train targets shape: {0}'.format(train_targets.shape))

        test_inputs, test_targets = batch["test"]
        print('Test inputs shape: {0}'.format(test_inputs.shape))
        print('Test targets shape: {0}'.format(test_targets.shape))
`

**Comment by iron-beagle** [comment]
# Issues
commit: 5922545

The sampler still has an inappropriate global RNG. The test below and its imported MetaDataset class uses the helpers that I made and not the base dataset classes. Deciding on what kind of transforms to use to get the images to the same shape should be left up to the user. This needs to be implemented rather than my simple resize put on all the helpers.

``` python

from torchvision import transforms
from torchmetal.datasets.metadata import MetaDataset
from torchmetal.datasets.helpers import omniglot, quickdraw, aircraft
from torchmetal.transforms import ClassSplitter, Categorical
from torchmetal.utils.data import BatchMetaDataLoader, MetaDataSampler
from torchmetal.datasets.metadataset.config.config import EpisodeDescriptionConfig

if __name__ == "__main__":
    episode_descr = EpisodeDescriptionConfig(
        num_ways=None,
        num_support=None,
        num_query=None,
        min_ways=5,
        max_ways_upper_bound=50,
        max_num_query=10,
        max_support_set_size=500,
        max_support_size_contrib_per_class=100,
        min_log_weight=np.log(0.5),
        max_log_weight=np.log(2),
        ignore_dag_ontology=False,
        ignore_bilevel_ontology=False,
        ignore_hierarchy_probability=0,
        simclr_episode_fraction=0,
    )

    dataset = MetaDataset(
        meta_train=True
    )

    sampler = MetaDataSampler(dataset, episode_descr)

    dataloader = BatchMetaDataLoader(
        dataset, shuffle=False, batch_size=16, num_workers=4, sampler=sampler
    )

    for batch in dataloader:
        train_inputs, train_targets = batch["train"]
        print(
            "Train inputs shape: {0}".format(train_inputs.shape)
        )  # (16, 500, 1, 28, 28)
        print("Train targets shape: {0}".format(train_targets.shape))  # (16, 500)

        test_inputs, test_targets = batch["test"]
        print(
            "Test inputs shape: {0}".format(test_inputs.shape)
        )  # (16, 500, 1, 28, 28)
        print("Test targets shape: {0}".format(test_targets.shape))  # (16, 500)
```
